@viirya `hadoop-client-check-invariants` failed:
```
[INFO] --- maven-dependency-plugin:3.0.2:build-classpath (put-client-artifacts-in-a-property) @ hadoop-client-check-invariants ---
[INFO] Dependencies classpath:
/home/jenkins/jenkins-agent/workspace/hadoop-multibranch_PR-3385/ubuntu-focal/src/hadoop-client-modules/hadoop-client-runtime/target/hadoop-client-runtime-3.4.0-SNAPSHOT.jar;/home/jenkins/jenkins-agent/workspace/hadoop-multibranch_PR-3385/ubuntu-focal/src/hadoop-client-modules/hadoop-client-api/target/hadoop-client-api-3.4.0-SNAPSHOT.jar
[INFO] 
[INFO] --- exec-maven-plugin:1.3.1:exec (check-jar-contents) @ hadoop-client-check-invariants ---
[ERROR] Found artifact with unexpected contents: '/home/jenkins/jenkins-agent/workspace/hadoop-multibranch_PR-3385/ubuntu-focal/src/hadoop-client-modules/hadoop-client-runtime/target/hadoop-client-runtime-3.4.0-SNAPSHOT.jar'
    Please check the following and either correct the build or update
    the allowed list with reasoning.

    org/xerial/
    org/xerial/snappy/
    org/xerial/snappy/BitShuffle.class
    org/xerial/snappy/BitShuffleNative.class
    org/xerial/snappy/BitShuffleType.class
    org/xerial/snappy/OSInfo.class
    org/xerial/snappy/PureJavaCrc32C.class
    org/xerial/snappy/Snappy.class
    org/xerial/snappy/SnappyApi.class
    org/xerial/snappy/SnappyBundleActivator.class
    org/xerial/snappy/SnappyCodec.class
    org/xerial/snappy/SnappyError.class
    org/xerial/snappy/SnappyErrorCode.class
    org/xerial/snappy/SnappyException.class
    org/xerial/snappy/SnappyFramed.class
    org/xerial/snappy/SnappyFramedInputStream$FrameAction.class
    org/xerial/snappy/SnappyFramedInputStream$FrameData.class
    org/xerial/snappy/SnappyFramedInputStream$FrameMetaData.class
    org/xerial/snappy/SnappyFramedInputStream.class
    org/xerial/snappy/SnappyFramedOutputStream.class
    org/xerial/snappy/SnappyHadoopCompatibleOutputStream.class
    org/xerial/snappy/SnappyIOException.class
    org/xerial/snappy/SnappyInputStream.class
    org/xerial/snappy/SnappyLoader.class
    org/xerial/snappy/SnappyNative.class
    org/xerial/snappy/SnappyOutputStream.class
    org/xerial/snappy/VERSION
    org/xerial/snappy/VERSION~
    org/xerial/snappy/buffer/
    org/xerial/snappy/buffer/BufferAllocator.class
    org/xerial/snappy/buffer/BufferAllocatorFactory.class
    org/xerial/snappy/buffer/CachedBufferAllocator$1.class
    org/xerial/snappy/buffer/CachedBufferAllocator.class
    org/xerial/snappy/buffer/DefaultBufferAllocator$1.class
    org/xerial/snappy/buffer/DefaultBufferAllocator.class
    org/xerial/snappy/native/
    org/xerial/snappy/native/AIX/
    org/xerial/snappy/native/AIX/ppc/
    org/xerial/snappy/native/AIX/ppc/libsnappyjava.a
    org/xerial/snappy/native/AIX/ppc64/
    org/xerial/snappy/native/AIX/ppc64/libsnappyjava.a
    org/xerial/snappy/native/FreeBSD/
    org/xerial/snappy/native/FreeBSD/x86_64/
    org/xerial/snappy/native/FreeBSD/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Linux/
    org/xerial/snappy/native/Linux/aarch64/
    org/xerial/snappy/native/Linux/aarch64/libsnappyjava.so
    org/xerial/snappy/native/Linux/android-arm/
    org/xerial/snappy/native/Linux/android-arm/libsnappyjava.so
    org/xerial/snappy/native/Linux/arm/
    org/xerial/snappy/native/Linux/arm/libsnappyjava.so
    org/xerial/snappy/native/Linux/armv6/
    org/xerial/snappy/native/Linux/armv6/libsnappyjava.so
    org/xerial/snappy/native/Linux/armv7/
    org/xerial/snappy/native/Linux/armv7/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc/
    org/xerial/snappy/native/Linux/ppc/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc64/
    org/xerial/snappy/native/Linux/ppc64/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc64le/
    org/xerial/snappy/native/Linux/ppc64le/libsnappyjava.so
    org/xerial/snappy/native/Linux/s390x/
    org/xerial/snappy/native/Linux/s390x/libsnappyjava.so
    org/xerial/snappy/native/Linux/x86/
    org/xerial/snappy/native/Linux/x86/libsnappyjava.so
    org/xerial/snappy/native/Linux/x86_64/
    org/xerial/snappy/native/Linux/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Mac/
    org/xerial/snappy/native/Mac/aarch64/
    org/xerial/snappy/native/Mac/aarch64/libsnappyjava.dylib
    org/xerial/snappy/native/Mac/x86/
    org/xerial/snappy/native/Mac/x86/libsnappyjava.jnilib
    org/xerial/snappy/native/Mac/x86_64/
    org/xerial/snappy/native/Mac/x86_64/libsnappyjava.dylib
    org/xerial/snappy/native/README
    org/xerial/snappy/native/SunOS/
    org/xerial/snappy/native/SunOS/sparc/
    org/xerial/snappy/native/SunOS/sparc/libsnappyjava.so
    org/xerial/snappy/native/SunOS/x86/
    org/xerial/snappy/native/SunOS/x86/libsnappyjava.so
    org/xerial/snappy/native/SunOS/x86_64/
    org/xerial/snappy/native/SunOS/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Windows/
    org/xerial/snappy/native/Windows/x86/
    org/xerial/snappy/native/Windows/x86/snappyjava.dll
    org/xerial/snappy/native/Windows/x86_64/
    org/xerial/snappy/native/Windows/x86_64/snappyjava.dll
    org/xerial/snappy/pool/
    org/xerial/snappy/pool/BufferPool.class
    org/xerial/snappy/pool/CachingBufferPool$1.class
    org/xerial/snappy/pool/CachingBufferPool$2.class
    org/xerial/snappy/pool/CachingBufferPool$IntFunction.class
    org/xerial/snappy/pool/CachingBufferPool.class
    org/xerial/snappy/pool/DefaultPoolFactory.class
    org/xerial/snappy/pool/DirectByteBuffers$1.class
    org/xerial/snappy/pool/DirectByteBuffers$2.class
    org/xerial/snappy/pool/DirectByteBuffers.class
    org/xerial/snappy/pool/QuiescentBufferPool.class
    org/xerial/snappy/pure/
    org/xerial/snappy/pure/PureJavaSnappy.class
    org/xerial/snappy/pure/SnappyConstants.class
    org/xerial/snappy/pure/SnappyRawCompressor.class
    org/xerial/snappy/pure/SnappyRawDecompressor.class
    org/xerial/snappy/pure/UnsafeUtil.class
[ERROR] Found artifact with unexpected contents: '/home/jenkins/jenkins-agent/workspace/hadoop-multibranch_PR-3385/ubuntu-focal/src/hadoop-client-modules/hadoop-client-api/target/hadoop-client-api-3.4.0-SNAPSHOT.jar'
    Please check the following and either correct the build or update
    the allowed list with reasoning.

    org/xerial/
    org/xerial/snappy/
    org/xerial/snappy/BitShuffle.class
    org/xerial/snappy/BitShuffleNative.class
    org/xerial/snappy/BitShuffleType.class
    org/xerial/snappy/OSInfo.class
    org/xerial/snappy/PureJavaCrc32C.class
    org/xerial/snappy/Snappy.class
    org/xerial/snappy/SnappyApi.class
    org/xerial/snappy/SnappyBundleActivator.class
    org/xerial/snappy/SnappyCodec.class
    org/xerial/snappy/SnappyError.class
    org/xerial/snappy/SnappyErrorCode.class
    org/xerial/snappy/SnappyException.class
    org/xerial/snappy/SnappyFramed.class
    org/xerial/snappy/SnappyFramedInputStream$FrameAction.class
    org/xerial/snappy/SnappyFramedInputStream$FrameData.class
    org/xerial/snappy/SnappyFramedInputStream$FrameMetaData.class
    org/xerial/snappy/SnappyFramedInputStream.class
    org/xerial/snappy/SnappyFramedOutputStream.class
    org/xerial/snappy/SnappyHadoopCompatibleOutputStream.class
    org/xerial/snappy/SnappyIOException.class
    org/xerial/snappy/SnappyInputStream.class
    org/xerial/snappy/SnappyLoader.class
    org/xerial/snappy/SnappyNative.class
    org/xerial/snappy/SnappyOutputStream.class
    org/xerial/snappy/VERSION
    org/xerial/snappy/VERSION~
    org/xerial/snappy/buffer/
    org/xerial/snappy/buffer/BufferAllocator.class
    org/xerial/snappy/buffer/BufferAllocatorFactory.class
    org/xerial/snappy/buffer/CachedBufferAllocator$1.class
    org/xerial/snappy/buffer/CachedBufferAllocator.class
    org/xerial/snappy/buffer/DefaultBufferAllocator$1.class
    org/xerial/snappy/buffer/DefaultBufferAllocator.class
    org/xerial/snappy/native/
    org/xerial/snappy/native/AIX/
    org/xerial/snappy/native/AIX/ppc/
    org/xerial/snappy/native/AIX/ppc/libsnappyjava.a
    org/xerial/snappy/native/AIX/ppc64/
    org/xerial/snappy/native/AIX/ppc64/libsnappyjava.a
    org/xerial/snappy/native/FreeBSD/
    org/xerial/snappy/native/FreeBSD/x86_64/
    org/xerial/snappy/native/FreeBSD/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Linux/
    org/xerial/snappy/native/Linux/aarch64/
    org/xerial/snappy/native/Linux/aarch64/libsnappyjava.so
    org/xerial/snappy/native/Linux/android-arm/
    org/xerial/snappy/native/Linux/android-arm/libsnappyjava.so
    org/xerial/snappy/native/Linux/arm/
    org/xerial/snappy/native/Linux/arm/libsnappyjava.so
    org/xerial/snappy/native/Linux/armv6/
    org/xerial/snappy/native/Linux/armv6/libsnappyjava.so
    org/xerial/snappy/native/Linux/armv7/
    org/xerial/snappy/native/Linux/armv7/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc/
    org/xerial/snappy/native/Linux/ppc/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc64/
    org/xerial/snappy/native/Linux/ppc64/libsnappyjava.so
    org/xerial/snappy/native/Linux/ppc64le/
    org/xerial/snappy/native/Linux/ppc64le/libsnappyjava.so
    org/xerial/snappy/native/Linux/s390x/
    org/xerial/snappy/native/Linux/s390x/libsnappyjava.so
    org/xerial/snappy/native/Linux/x86/
    org/xerial/snappy/native/Linux/x86/libsnappyjava.so
    org/xerial/snappy/native/Linux/x86_64/
    org/xerial/snappy/native/Linux/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Mac/
    org/xerial/snappy/native/Mac/aarch64/
    org/xerial/snappy/native/Mac/aarch64/libsnappyjava.dylib
    org/xerial/snappy/native/Mac/x86/
    org/xerial/snappy/native/Mac/x86/libsnappyjava.jnilib
    org/xerial/snappy/native/Mac/x86_64/
    org/xerial/snappy/native/Mac/x86_64/libsnappyjava.dylib
    org/xerial/snappy/native/README
    org/xerial/snappy/native/SunOS/
    org/xerial/snappy/native/SunOS/sparc/
    org/xerial/snappy/native/SunOS/sparc/libsnappyjava.so
    org/xerial/snappy/native/SunOS/x86/
    org/xerial/snappy/native/SunOS/x86/libsnappyjava.so
    org/xerial/snappy/native/SunOS/x86_64/
    org/xerial/snappy/native/SunOS/x86_64/libsnappyjava.so
    org/xerial/snappy/native/Windows/
    org/xerial/snappy/native/Windows/x86/
    org/xerial/snappy/native/Windows/x86/snappyjava.dll
    org/xerial/snappy/native/Windows/x86_64/
    org/xerial/snappy/native/Windows/x86_64/snappyjava.dll
    org/xerial/snappy/pool/
    org/xerial/snappy/pool/BufferPool.class
    org/xerial/snappy/pool/CachingBufferPool$1.class
    org/xerial/snappy/pool/CachingBufferPool$2.class
    org/xerial/snappy/pool/CachingBufferPool$IntFunction.class
    org/xerial/snappy/pool/CachingBufferPool.class
    org/xerial/snappy/pool/DefaultPoolFactory.class
    org/xerial/snappy/pool/DirectByteBuffers$1.class
    org/xerial/snappy/pool/DirectByteBuffers$2.class
    org/xerial/snappy/pool/DirectByteBuffers.class
    org/xerial/snappy/pool/QuiescentBufferPool.class
    org/xerial/snappy/pure/
    org/xerial/snappy/pure/PureJavaSnappy.class
    org/xerial/snappy/pure/SnappyConstants.class
    org/xerial/snappy/pure/SnappyRawCompressor.class
    org/xerial/snappy/pure/SnappyRawDecompressor.class
    org/xerial/snappy/pure/UnsafeUtil.class
```

you'll need to check the build artifacts for the log

--- a/hadoop-client-modules/hadoop-client-check-invariants/src/test/resources/ensure-jars-have-correct-contents.sh
+++ b/hadoop-client-modules/hadoop-client-check-invariants/src/test/resources/ensure-jars-have-correct-contents.sh
@@ -67,6 +67,8 @@ allowed_expr+="|^krb5_udp-template.conf$"
 # Jetty uses this style sheet for directory listings. TODO ensure our
 # internal use of jetty disallows directory listings and remove this.
 allowed_expr+="|^jetty-dir.css$"
+# Snappy java is native library. We cannot relocate it to under org/apache/hadoop.
+allowed_expr+="|^org/xerial/"
 
 allowed_expr+=")"
 declare -i bad_artifacts=0

--- a/hadoop-client-modules/hadoop-client-integration-tests/pom.xml
+++ b/hadoop-client-modules/hadoop-client-integration-tests/pom.xml
@@ -52,6 +52,11 @@
       <artifactId>junit</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.lz4</groupId>
+      <artifactId>lz4-java</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
   <profiles>
     <profile>

--- a/hadoop-client-modules/hadoop-client-integration-tests/src/test/java/org/apache/hadoop/example/ITUseHadoopCodecs.java
+++ b/hadoop-client-modules/hadoop-client-integration-tests/src/test/java/org/apache/hadoop/example/ITUseHadoopCodecs.java
@@ -1,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.hadoop.example;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+import java.io.*;
+import java.util.Random;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeys;
+import org.apache.hadoop.io.DataInputBuffer;
+import org.apache.hadoop.io.DataOutputBuffer;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.hadoop.io.RandomDatum;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.CompressionInputStream;
+import org.apache.hadoop.io.compress.CompressionOutputStream;
+import org.apache.hadoop.io.compress.zlib.ZlibFactory;
+import org.apache.hadoop.util.ReflectionUtils;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Ensure that we can perform codec operations against the shaded minicluster
+ * given the API and runtime jars by performing some simple smoke tests.
+ */
+public class ITUseHadoopCodecs {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ITUseHadoopCodecs.class);
+
+  private Configuration conf = new Configuration();
+  private int count = 100;
+  private int seed = new Random().nextInt();
+
+  @Test
+  public void testGzipCodec() throws IOException {
+    ZlibFactory.setNativeZlibLoaded(false);
+    assertFalse(ZlibFactory.isNativeZlibLoaded(conf));
+    codecTest(conf, seed, 0, "org.apache.hadoop.io.compress.GzipCodec");
+    codecTest(conf, seed, count, "org.apache.hadoop.io.compress.GzipCodec");
+  }
+
+  @Test
+  public void testSnappyCodec() throws IOException {
+    codecTest(conf, seed, 0, "org.apache.hadoop.io.compress.SnappyCodec");
+    codecTest(conf, seed, count, "org.apache.hadoop.io.compress.SnappyCodec");
+  }
+
+  @Test
+  public void testLz4Codec() throws IOException {
+    conf.setBoolean(
+      CommonConfigurationKeys.IO_COMPRESSION_CODEC_LZ4_USELZ4HC_KEY,
+      false);
+    codecTest(conf, seed, 0, "org.apache.hadoop.io.compress.Lz4Codec");
+    codecTest(conf, seed, count, "org.apache.hadoop.io.compress.Lz4Codec");
+    conf.setBoolean(
+      CommonConfigurationKeys.IO_COMPRESSION_CODEC_LZ4_USELZ4HC_KEY,
+      true);
+    codecTest(conf, seed, 0, "org.apache.hadoop.io.compress.Lz4Codec");
+    codecTest(conf, seed, count, "org.apache.hadoop.io.compress.Lz4Codec");
+  }
+
+  private void codecTest(Configuration conf, int seed, int count, String codecClass)
+    throws IOException {
+
+    // Create the codec
+    CompressionCodec codec = null;
+    try {
+      codec = (CompressionCodec)
+              ReflectionUtils.newInstance(conf.getClassByName(codecClass), conf);
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException("Illegal codec!");
+    }
+    LOG.info("Created a Codec object of type: " + codecClass);
+
+    // Generate data
+    DataOutputBuffer data = new DataOutputBuffer();
+    RandomDatum.Generator generator = new RandomDatum.Generator(seed);
+    for(int i = 0; i < count; ++i) {
+      generator.next();
+      RandomDatum key = generator.getKey();
+      RandomDatum value = generator.getValue();
+
+      key.write(data);
+      value.write(data);
+    }
+    LOG.info("Generated " + count + " records");
+
+    // Compress data
+    DataOutputBuffer compressedDataBuffer = new DataOutputBuffer();
+    try (CompressionOutputStream deflateFilter =
+      codec.createOutputStream(compressedDataBuffer);
+      DataOutputStream deflateOut =
+        new DataOutputStream(new BufferedOutputStream(deflateFilter))) {
+      deflateOut.write(data.getData(), 0, data.getLength());
+      deflateOut.flush();
+      deflateFilter.finish();
+    }
+
+    // De-compress data
+    DataInputBuffer deCompressedDataBuffer = new DataInputBuffer();
+    deCompressedDataBuffer.reset(compressedDataBuffer.getData(), 0,
+            compressedDataBuffer.getLength());
+    DataInputBuffer originalData = new DataInputBuffer();
+    originalData.reset(data.getData(), 0, data.getLength());
+    try (CompressionInputStream inflateFilter =
+      codec.createInputStream(deCompressedDataBuffer);
+      DataInputStream originalIn =
+        new DataInputStream(new BufferedInputStream(originalData))) {
+
+      // Check
+      int expected;
+      do {
+        expected = originalIn.read();
+        assertEquals("Inflated stream read by byte does not match",
+                expected, inflateFilter.read());
+      } while (expected != -1);
+    }
+
+    LOG.info("SUCCESS! Completed checking " + count + " records");
+  }
+}

